File: MiniProject/PageManager.java
package miniProject;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Map;
import java.util.Scanner;
import miniProject.StateName.PageName;
import miniProject.StateName.UserType;

public class PageManager {

	private SingletonState inst;
	private UserManager user;
	private PageName page;
	private Scanner scanner = new Scanner(System.in);

	public PageManager(SingletonState inst) {
		this.inst = inst;
		user = new UserManager();
		this.page = inst.currentPage;
	}

	public void startProgram() {
		page = PageName.LOGIN_PAGE;
		System.out.println("What do you want to do?\n 1:Buy item(User)\n 2:Manage item(Admin)\n 3:Exit");
		int choice = scanner.nextInt();
		 
		if (choice == 1) {
			user.setType(UserType.USER);
			mainUserPage(user);
		} else if (choice == 2) {
			user.setType(UserType.ADMIN);
			mainAdminPage(user);
		} else if (choice == 3) {
			System.out.println("Thank you!");
			return;
		} else {
			System.out.println("\nFalse Input!\n");
			startProgram();
		}
	}

	public void mainUserPage(UserManager usr) {
		page = PageName.USER_PAGE;
		int choice = 0;
		System.out.println("Main Page");
		System.out.println("Product List");
		inst.getProductManager().showAll();
		System.out.println("What would you like to do?");
		System.out.println(" 1:Add product to Cart");
		System.out.println(" 2:Watch products in cart");
		System.out.println(" 3:Logout");
		choice = scanner.nextInt();
		 
		if (choice == 1) {
			addToCartPage(usr);
		} else if (choice == 2) {
			showCartPage(user);
		} else if (choice == 3) {
			back();
		} else {
			System.out.println("\nWrong Input!\n");
			mainUserPage(user);
		}
	}

	public void mainAdminPage(UserManager usr) {
		page = PageName.ADMIN_PAGE;
		int choice = 0;
		System.out.println("\nHello " + usr.getType());
		System.out.println("Product List");
		inst.getProductManager().showAll();
		System.out.println("What would you like to do?");
		System.out.println(" 1:Add Product");
		System.out.println(" 2:Remove Product");
		System.out.println(" 3:Logout");
		choice = scanner.nextInt();
		 
		if (choice == 1) {
			addProductToList();
		} else if (choice == 2) {
			removeProductfromList();
		} else if (choice == 3) {
			back();
		} else {
			System.out.println("\nWrong Input!\n");
			mainAdminPage(user);
		}
	}

	public void showCartPage(UserManager usr) {
		page = PageName.CART_PAGE;
		System.out.println("Your Cart:");
		System.out.println(usr.watchCart());
		if (usr.cart.isEmpty()) {
			back();
		}
		System.out.println("What would you like to do?");
		System.out.println(" 1:Delete item in the cart");
		System.out.println(" 2:Check out");
		System.out.println(" 3:Back");
		int choice = scanner.nextInt();
		 
		if (choice == 1) {
			removeFromCartPage();
		} else if (choice == 2) {
			SimpleDateFormat formatter = new SimpleDateFormat("yyyy-MM-dd 'at' HH:mm:ss z");
			Date date = new Date(System.currentTimeMillis());
			String bill = "Order receipt:\n" + usr.watchCart() + "\n" + formatter.format(date) + "\nAddress:\n";
			checkOutPage(bill);
		} else if (choice == 3) {
			back();
		} else {
			System.out.println("Wrong input!");
			showCartPage(user);
		}
	}

	public void addToCartPage(UserManager usr) {
		page = PageName.USER_ADD_PAGE;
		System.out.println("Product List");
		inst.getProductManager().showAll();
		System.out.println("What would you like to add?");
		boolean foundedProduct = false;
		boolean foundedinCart = false;
		int choice = scanner.nextInt();
		 
		for (Product product : inst.getProductManager().products) {
			if (choice == product.getID()) {
				for (Map.Entry<Product, Integer> set : user.cart.entrySet()) {
					if (set.getKey().name.equals(product.name)) {
						foundedinCart = true;
						set.setValue(set.getValue() + 1);
					}
				}
				if (!foundedinCart) {
					user.cart.put(product, 1);
				}
				System.out.println("Added " + product.name + " to cart");
				foundedProduct = true;
				break;
			}
		}
		if (!foundedProduct) {
			System.out.println("Not Found this item!");
		}
		back();
	}

	public void removeFromCartPage() {
		page = PageName.USER_REMOVE_PAGE;
		System.out.println("What product that you want to remove from your cart?");
		for (Map.Entry<Product, Integer> set : user.cart.entrySet()) {
			System.out.println(" " + set.getKey().toString());
		}
		int choice = scanner.nextInt();
		 
		String removed = user.deleteProductInCart(choice);
		if (removed == null) {
			System.out.println("Not has this product in Cart!");
		} else {
			System.out.println("Removed " + removed + " from your cart.\n");
		}
		back();
	}

	public void checkOutPage(String bill) {
		System.out.println("Please add your address for shipping : ");
		Scanner addrScanner = new Scanner(System.in);
		String addr = addrScanner.nextLine();
		try {
			SimpleDateFormat formatter = new SimpleDateFormat("yyyyMMddHHmmss");
			Date date = new Date(System.currentTimeMillis());
			String fileName = "Receipt_" + formatter.format(date) + ".txt";
			FileWriter myWriter = new FileWriter(fileName);
			myWriter.write(bill + addr);
			myWriter.close();
			System.out.println("Successfully saved the receipt.");
		} catch (IOException e) {
			System.out.println("An error occurred.");
			e.printStackTrace();
		}
		user.cart.clear();
		mainUserPage(user);
	}

	public void addProductToList() {
		page = PageName.ADMIN_ADD_PAGE;
		System.out.println("Please enter product name to add : ");
		String productName = scanner.next();
		System.out.println("Please enter price : ");
		int productPrice = scanner.nextInt();
		 
		Product newProduct = new Product(productName, productPrice);
		inst.productManager.addProduct(newProduct);
		System.out.println("Add " + productName + " to Productlist!");
		System.out.println();
		System.out.println("Product List");
		for (Product product : inst.getProductManager().products) {
			System.out.println(" (" + product.getID() + ")" + product.name);
		}
		System.out.println();
		back();
	}

	public void removeProductfromList() {
		page = PageName.ADMIN_REMOVE_PAGE;
		System.out.println("Product List");
		inst.getProductManager().showAll();
		System.out.println("Please fill product id that you want to remove from list?");
		int idProductRemove = scanner.nextInt();
		 
		inst.productManager.removeProductbyID(idProductRemove);
		System.out.println("Product List");
		for (Product product : inst.getProductManager().products) {
			System.out.println(" (" + product.getID() + ")" + product.name);
		}
		back();
	}

	public void back() {
		switch (page) {
		case ADMIN_PAGE:
			startProgram();
			break;
		case USER_PAGE:
			startProgram();
			break;
		case CART_PAGE:
			mainUserPage(user);
			break;
		case USER_REMOVE_PAGE:
			showCartPage(user);
			break;
		case USER_ADD_PAGE:
			mainUserPage(user);
			break;
		case ADMIN_REMOVE_PAGE:
			mainAdminPage(user);
			break;
		case ADMIN_ADD_PAGE:
			mainAdminPage(user);
			break;
		default:
			break;
		}
	}

}

File: MiniToTypeScript/pageManager.js
"use strict";
exports.__esModule = true;
exports.PageManager = void 0;
var product_js_1 = require("./product.js");
var myEnum_js_1 = require("./myEnum.js");
var userManager_js_1 = require("./userManager.js");
//import * as readline from 'readline'
var PageManager = /** @class */ (function () {
    function PageManager(inst) {
        this.inst = inst;
        this.user = new userManager_js_1.UserManager();
        this.page = inst.currentPage;
    }
    PageManager.prototype.startProgram = function () {
        this.page = myEnum_js_1.PageName.LOGIN_PAGE;
        console.log("What do you want to do?\n 1:Buy item(User)\n 2:Manage item(Admin)\n 3:Exit");
        var choice;
        var readline = require('readline');
        var rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
        rl.on('line', function (input) {
            choice = parseInt(input);
        });
        rl.close();
        if (choice == 1) {
            this.user.setType(myEnum_js_1.UserType.USER);
            this.mainUserPage(this.user);
        }
        else if (choice == 2) {
            this.user.setType(myEnum_js_1.UserType.ADMIN);
            this.mainAdminPage(this.user);
        }
        else if (choice == 3) {
            console.log("Thank you!");
            return;
        }
        else {
            console.log("\nFalse Input!\n");
            this.startProgram();
        }
    };
    PageManager.prototype.mainUserPage = function (usr) {
        this.page = myEnum_js_1.PageName.USER_PAGE;
        var choice = 0;
        console.log("Main Page");
        console.log("Product List");
        this.inst.getProductManager().showAll();
        console.log("What would you like to do?");
        console.log(" 1:Add product to Cart");
        console.log(" 2:Watch products in cart");
        console.log(" 3:Logout");
        var readline = require('readline');
        var rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
        rl.on('line', function (input) {
            choice = parseInt(input);
        });
        rl.close();
        if (choice == 1) {
            this.addToCartPage(usr);
        }
        else if (choice == 2) {
            this.showCartPage(usr);
        }
        else if (choice == 3) {
            this.back();
        }
        else {
            console.log("\nWrong Input!\n");
            this.mainUserPage(usr);
        }
    };
    PageManager.prototype.mainAdminPage = function (usr) {
        this.page = myEnum_js_1.PageName.ADMIN_PAGE;
        var choice = 0;
        console.log("\nHello " + usr.getType());
        console.log("Product List");
        this.inst.getProductManager().showAll();
        console.log("What would you like to do?");
        console.log(" 1:Add Product");
        console.log(" 2:Remove Product");
        console.log(" 3:Logout");
        var readline = require('readline');
        var rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
        rl.on('line', function (input) {
            choice = parseInt(input);
        });
        rl.close();
        if (choice == 1) {
            this.addProductToList();
        }
        else if (choice == 2) {
            this.removeProductfromList();
        }
        else if (choice == 3) {
            this.back();
        }
        else {
            console.log("\nWrong Input!\n");
            this.mainAdminPage(this.user);
        }
    };
    PageManager.prototype.showCartPage = function (usr) {
        this.page = myEnum_js_1.PageName.CART_PAGE;
        console.log("Your Cart:");
        console.log(usr.watchCart());
        if (usr.cart.length <= 0) {
            this.back();
        }
        console.log("What would you like to do?");
        console.log(" 1:Delete item in the cart");
        console.log(" 2:Check out");
        console.log(" 3:Back");
        var choice;
        var readline = require('readline');
        var rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
        rl.on('line', function (input) {
            choice = parseInt(input);
        });
        rl.close();
        if (choice == 1) {
            this.removeFromCartPage();
        }
        else if (choice == 2) {
            var date = new Date();
            var dd = date.getDate().toString();
            var MM = (date.getUTCMonth() + 1).toString();
            var yyyy = date.getFullYear().toString();
            var HH = date.getHours().toString();
            var mm = date.getMinutes().toString();
            var ss = date.getSeconds().toString();
            var bill = "Order receipt:\n" + usr.watchCart() + "\n" + (yyyy + "-" + mm + "-" + dd + " at " + HH + ":" + mm + ":" + ss) + "\nAddress:\n";
            this.checkOutPage(bill, "" + yyyy + MM + dd + HH + mm + ss);
        }
        else if (choice == 3) {
            this.back();
        }
        else {
            console.log("Wrong input!");
            this.showCartPage(this.user);
        }
    };
    PageManager.prototype.addToCartPage = function (usr) {
        this.page = myEnum_js_1.PageName.USER_ADD_PAGE;
        console.log("Product List");
        this.inst.getProductManager().showAll();
        console.log("What would you like to add?");
        var foundedProduct = false;
        var foundedinCart = false;
        var choice;
        var readline = require('readline');
        var rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
        rl.on('line', function (input) {
            choice = parseInt(input);
        });
        rl.close();
        if (!usr.addProductToCartbyID(choice, 1)) {
            console.log("Not Found this item!");
        }
        this.back();
    };
    PageManager.prototype.removeFromCartPage = function () {
        this.page = myEnum_js_1.PageName.USER_REMOVE_PAGE;
        console.log("What product that you want to remove from your cart?");
        for (var _i = 0, _a = this.user.cart; _i < _a.length; _i++) {
            var set = _a[_i];
            console.log(" " + set["product"].toString());
        }
        var choice;
        var readline = require('readline');
        var rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
        rl.on('line', function (input) {
            choice = parseInt(input);
        });
        rl.close();
        var removed = this.user.deleteProductInCartI(choice);
        if (removed == null) {
            console.log("Not has this product in Cart!");
        }
        else {
            console.log("Removed " + removed + " from your cart.\n");
        }
        this.back();
    };
    PageManager.prototype.checkOutPage = function (bill, d) {
        console.log("Please add your address for shipping : ");
        var readline = require('readline');
        var rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
        var addr;
        rl.on('line', function (input) {
            addr = input;
        });
        rl.close();
        try {
            var fileName = "Receipt_" + d + ".txt";
            var fs = require('fs');
            fs.writeFile(fileName, bill + addr, 'utf8');
            console.log("Successfully saved the receipt.");
        }
        catch (exception) {
            console.log("An error occurred.");
        }
        this.user.cart = [];
        this.mainUserPage(this.user);
    };
    PageManager.prototype.addProductToList = function () {
        this.page = myEnum_js_1.PageName.ADMIN_ADD_PAGE;
        var readline = require('readline');
        var rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
        console.log("Please enter product name to add : ");
        var productName;
        rl.on('line', function (input) {
            productName = input;
        });
        console.log("Please enter price : ");
        var productPrice;
        rl.on('line', function (input) {
            productPrice = parseFloat(input);
        });
        rl.close();
        var newProduct = new product_js_1.Product(productName, productPrice);
        this.inst.productManager.addProduct(newProduct);
        console.log("Add " + productName + " to Productlist!\n");
        console.log("Product List");
        for (var _i = 0, _a = this.inst.getProductManager().products; _i < _a.length; _i++) {
            var product = _a[_i];
            console.log(" (" + product.getID() + ")" + product.name);
        }
        console.log();
        this.back();
    };
    PageManager.prototype.removeProductfromList = function () {
        this.page = myEnum_js_1.PageName.ADMIN_REMOVE_PAGE;
        console.log("Product List");
        this.inst.getProductManager().showAll();
        console.log("Please fill product id that you want to remove from list?");
        var idProductRemove;
        var readline = require('readline');
        var rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
        rl.on('line', function (input) {
            idProductRemove = parseInt(input);
        });
        rl.close();
        this.inst.productManager.removeProductbyID(idProductRemove);
        console.log("Product List");
        for (var _i = 0, _a = this.inst.getProductManager().products; _i < _a.length; _i++) {
            var product = _a[_i];
            console.log(" (" + product.getID() + ")" + product.name);
        }
        this.back();
    };
    PageManager.prototype.back = function () {
        switch (this.page) {
            case myEnum_js_1.PageName.ADMIN_PAGE:
                this.startProgram();
                break;
            case myEnum_js_1.PageName.USER_PAGE:
                this.startProgram();
                break;
            case myEnum_js_1.PageName.CART_PAGE:
                this.mainUserPage(this.user);
                break;
            case myEnum_js_1.PageName.USER_REMOVE_PAGE:
                this.showCartPage(this.user);
                break;
            case myEnum_js_1.PageName.USER_ADD_PAGE:
                this.mainUserPage(this.user);
                break;
            case myEnum_js_1.PageName.ADMIN_REMOVE_PAGE:
                this.mainAdminPage(this.user);
                break;
            case myEnum_js_1.PageName.ADMIN_ADD_PAGE:
                this.mainAdminPage(this.user);
                break;
            default:
                break;
        }
    };
    return PageManager;
}());
exports.PageManager = PageManager;
